File: solutions/ruby/protein-translation/protein_translation.rb
class Translation
  STOP_CODON = "STOP".freeze
  private_constant :STOP_CODON

  def self.of_rna(strand)
    proteins = []
    chunk_every(strand, 3).each do |chunk|
      protein = of_codon(chunk)
      break if protein == STOP_CODON
      proteins.append(protein)
    end
    return proteins
  end

  private

  def self.of_codon(codon)
    case codon
    when "AUG"
      return "Methionine"
    when "UUU", "UUC"
      return "Phenylalanine"
    when "UUA", "UUG"
      return "Leucine"
    when "UCU", "UCC", "UCA", "UCG"
      return "Serine"
    when "UAU", "UAC"
      return "Tyrosine"
    when "UGU", "UGC"
      return "Cysteine"
    when "UGG"
      return "Tryptophan"
    when "UAA", "UAG", "UGA"
      return STOP_CODON
    else
      raise InvalidCodonError
    end
  end

  def self.chunk_every(string, size)
    (0..(string.length - 1) / size).map { |i| string[i * size, size] }
  end
end

class InvalidCodonError < StandardError
end

File: solutions/erlang/protein-translation/src/protein_translation.erl
-module(protein_translation).

-export([proteins/1]).

-define(PROTEINS, [
    {"AUG", methionine},
    {"UGG", tryptophan},
    {"UUU", phenylalanine},
    {"UUC", phenylalanine},
    {"UUA", leucine},
    {"UUG", leucine},
    {"UCU", serine},
    {"UCC", serine},
    {"UCA", serine},
    {"UCG", serine},
    {"UAU", tyrosine},
    {"UAC", tyrosine},
    {"UGU", cysteine},
    {"UGC", cysteine},
    {"UAA", stopCodon},
    {"UAG", stopCodon},
    {"UGA", stopCodon}
]).

proteins(_Strand) ->
    from_codons(_Strand, []).

from_codons([X, Y, Z | Rest], Acc) ->
    case proplists:get_value([X, Y, Z], ?PROTEINS) of
        stopCodon -> from_codons([], Acc);
        undefined -> {error, badarg};
        Protein -> from_codons(Rest, [Protein | Acc])
    end;
from_codons([], Acc) ->
    {ok, lists:reverse(Acc)};
from_codons(_, _) ->
    {error, badarg}.
